core/src/main/scala/com/salesforce/op/stages/impl/tuning/DataCutter.scala

```diff
@@ -82,33 +85,87 @@ class DataCutter(uid: String = UID[DataCutter]) extends Splitter(uid = uid) with
    * @param data
    * @return Parameters set in examining data
    */
-  override def preValidationPrepare(data: Dataset[Row]): Option[SplitterSummary] = {
-
-    import data.sparkSession.implicits._
-
-    val labels = data.map(r => r.getDouble(0) -> 1L)
-    val labelCounts = labels.groupBy(labels.columns(0)).sum(labels.columns(1)).persist()
-    val (resKeep, resDrop) = estimate(labelCounts)
-    labelCounts.unpersist()
-    setLabels(resKeep, resDrop)
+  override def preValidationPrepare(data: Dataset[Row], labelColNameOpt: Option[String]): Option[SplitterSummary] = {
+    val labelColName = labelColNameOpt.getOrElse(data.columns(0))
+    val labelColIdx = data.columns.indexOf(labelColName)
+
+    if (Try(getLabelsToKeep).toOption.isEmpty) {
+      val labelCounts = data.groupBy(labelColName).count().persist()
+      val res = estimate(labelCounts)
+      labelCounts.unpersist()
+      setLabels(res.labelsKept.distinct, res.labelsDropped.distinct, res.labelsDroppedTotal)
+    }
 
-    summary = Option(DataCutterSummary(labelsKept = getLabelsToKeep, labelsDropped = getLabelsToDrop))
+    val newSummary = summary
+      .collect {
+        case dcs: DataCutterSummary if dcs.preparedDF.isDefined => dcs
+      }
+      .getOrElse {
+        val labelMetaArr = getLabelsFromMetadata(data)
+        val labelSet = getLabelsToKeep.toSet
+
+        val dataPrep = {
+          log.info(s"Dropping rows with columns not in $labelSet")
+
+          // Update metadata that spark.ml Classifier is using tp determine the number of classes
+          //
+          val na = NominalAttribute.defaultAttr.withName(labelColName)
+          val metadataNA = if (labelMetaArr.isEmpty) {
+            na
+          } else {
+            na.withValues(labelMetaArr.take(labelSet.size))
```

Comment:
I think part of the problem may be the assumption that the first labels are the ones that you keep (they should be but there are probably edge cases where it is not true). `labelMetaArr.zipWithIndex.filter{(case (_, i))labelSet.contains(i)}`

Reply:
agreed, more robust and elegant